jax/_src/numpy/util.py

```diff
@@ -117,6 +119,44 @@ def _parse_extra_params(extra_params: str) -> dict[str, str]:
   return {p.partition(' : ')[0].partition(', ')[0]: p for p in parameters}
 
 
+def _get_device_set(x: ArrayLike | xc.Device | Sharding | None) -> Set[xc.Device] | None:
+  if x is None or isinstance(x, core.Tracer):
+    return None
+  elif isinstance(x, Sharding):
+    return x.device_set
+  elif isinstance(x, xc.Device):
+    return {x}
+  elif hasattr(x, "devices"):
+    return x.devices()
+  else:
+    raise ValueError(f"Attempted to get a device set from an unsupported type: {type(x)}")
+
+
+def _place_array(x: Array, device: xc.Device | Sharding | None = None, copy=None) -> Array:
+  """Helper utility for copying an array, or placing it on a device or sharding.
+
+  Note that `device_put` is regarded as a no-op under JIT compilation, so we
+  ensure array API device semantics compliance only under eager execution,
+  favoring JIT compilation performance over correctness in this case.
+  """
+
+  devices = _get_device_set(device)
+  src_devices = _get_device_set(x)
```

Comment:
I think instead of a function, we should replace this with a handler based system. So you register handlers for array types and they tell you what you get. It can be a device or a sharding. For a jax.Array, it should always be a sharding.